### PR TITLE
impl(common): support extracting a typed option from an Options

### PR DESCRIPTION
**Pull request by devbww**
```
absl::optional<typename T::Type> internal::ExtractOption<T>(Options&);
```
Extracts and returns the value for `T` from the options, or nullopt if `T` was not set.

This is intended for code paths that want to consume an option, for example in a client call that passes the option's value via conventional mechanisms, rather than through an `OptionsSpan`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11576)

### CODE REVIEW

File: google/cloud/options_test.cc [hunk 1]
```diff
@@ -270,6 +270,30 @@ TEST(MergeOptions, Basics) {
...
+  EXPECT_TRUE(opts.has<IntOption>());
+  EXPECT_TRUE(opts.has<StringOption>());
+  EXPECT_FALSE(b.has_value());
+
+  auto i = internal::ExtractOption<IntOption>(opts);
+  EXPECT_FALSE(opts.has<BoolOption>());
+  EXPECT_FALSE(opts.has<IntOption>());
+  EXPECT_TRUE(opts.has<StringOption>());
+  ASSERT_TRUE(i.has_value());
+  EXPECT_EQ(*i, 42);
```
Comment (dbolduc) [inline google/cloud/options_test.cc:287]: nit: `EXPECT_THAT(i, Optional(42));`

File: google/cloud/options_test.cc [hunk 2]
```diff
@@ -270,6 +270,30 @@ TEST(MergeOptions, Basics) {
...
+
+  auto s = internal::ExtractOption<StringOption>(opts);
+  EXPECT_FALSE(opts.has<BoolOption>());
+  EXPECT_FALSE(opts.has<IntOption>());
+  EXPECT_FALSE(opts.has<StringOption>());
+  ASSERT_TRUE(s.has_value());
+  EXPECT_EQ(*s, "foo");
```
Comment (dbolduc) [inline google/cloud/options_test.cc:294]: nit: `EXPECT_THAT(i, Optional("foo"));`

File: google/cloud/options.h [hunk 3]
```diff
@@ -320,6 +325,22 @@ void CheckExpectedOptions(Options const& opts, char const* caller) {
  */
 Options MergeOptions(Options preferred, Options alternatives);
 
+/**
+ * Extracts and returns the value for `T` from @p opts, or nullopt if `T` was
+ * not set.  This is intended for code paths that want to consume an option,
+ * for example, a client call that passes the option's value via conventional
+ * mechanisms, rather than though an OptionsSpan.
```
Comment (dbolduc) [inline google/cloud/options.h:332]: s/though/through/

**Review by dbolduc — APPROVED** [review]

File: google/cloud/options_test.cc (hunk 1 above)
Comment (devbww) [reply google/cloud/options_test.cc]: Done.

File: google/cloud/options_test.cc (hunk 2 above)
Comment (devbww) [reply google/cloud/options_test.cc]: Done.

File: google/cloud/options.h (hunk 3 above)
Comment (devbww) [reply google/cloud/options.h]: Done.

**Pull request merged**